crates/runtime/src/logic.rs

```diff
@@ -426,4 +484,36 @@ impl VMHostFunctions<'_> {
         self.with_logic_mut(|logic| logic.registers.set(logic.limits, register_id, data))?;
         Ok(status)
     }
+
+    /// Generates a new UUID using v4 standard.
+    ///
+    /// This function generates a new UUID by calling the host. The UUID is
+    /// dependent on randomness, which is not available inside the guest
+    /// runtime. Therefore the guest needs to request this from the host.
+    ///
+    pub fn generate_uuid(&mut self, register_id: u64) -> VMLogicResult<()> {
+        let uuid = self.borrow_logic().generate_uuid();
+        self.with_logic_mut(|logic| {
+            logic
+                .registers
+                .set(logic.limits, register_id, *uuid.as_bytes())
+        })?;
+        Ok(())
+    }
+
+    /// Gets the current time.
+    ///
+    /// This function obtains the current time as a nanosecond timestamp, as
+    /// [`SystemTime`] is not available inside the guest runtime. Therefore the
+    /// guest needs to request this from the host.
+    ///
+    pub fn time_now(&mut self, register_id: u64) -> VMLogicResult<()> {
+        let now = self.borrow_logic().time_now();
```

Comment:
same here

```suggestion
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .expect("Time went backwards to before the Unix epoch!")
            .as_nanos() as u64
```

Reply:
Agreed - calling `logic.time_now()` here is better. Additionally, in this location we can indeed subsume the `logic.time_now()` call into this location, as it's the only time we use it. So although it was added to try to preserve consistency with the other, similar implementations, there is no special need for a separate function here. So this has been removed 🙂 👍

Reply:
> Additionally, in this location we can indeed subsume the logic.time_now() call into this location

not necessarily, I like to think of these like `unsafe {}` (it technically is behind the scenes too), limit the scope of the closure to the use of `logic` and time doesn't use `logic`, so it can be outside

the same way unsafe things should be outside unsafe blocks even though unsafe things use their outputs

Reply:
> > Additionally, in this location we can indeed subsume the logic.time_now() call into this location
> 
> not necessarily

I am confused - this was your suggestion? 😄 I was agreeing with you that this could indeed happen here, and I've made that change.

If the change is incorrect, can you explain what you intend more clearly?

If the focus is about the logic, then the move inside agrees with your comments.

If the focus is about what's unsafe or not, then that's the way I had previously approached it, in both locations.

Reply:
> If the change is incorrect
> If the focus is about the logic, then the move inside agrees with your comments.

so `logic` is just a variable, `with_logic_mut` is a safe way of fetching it because it's self-referencing and that requires `unsafe`

so it's not incorrect, it just makes sense to limit the scope of the closure to things that _actually need to use the variable_

Reply:
Localizing the call to this location without involving `VMLogic`, yes.

However, the `let now = _` statement should be outside the closure.

Reply:
Thank you for confirming 🙂